### PR TITLE
[Performance][SFA] Pack grouped full-visible prefill rows into FIA

### PR DESCRIPTION
**Pull request by Ruiqiu-Zheng**
### What this PR does / why we need it?

This PR adds a default-off grouped fast path for the shared SFA consumer during prefill. It does not change LightningIndexer producer ownership or TopK publication.

For each request, the consumer identifies the query-row prefix whose visible KV window is still fully covered by `topk=2048`. Across the batch, the current first version enables the fast path only when those eligible prefixes pack to exactly 2048 dense query rows **and** a non-empty sparse tail remains. The packed rows use fused infer attention (FIA); the tail continues to consume the already-resolved `topk_indices` through the existing sparse SFA path. Results are restored to their original request-row order.

The admission is intentionally fail-closed. Low-fill, whole-dense, over-cap, unsupported geometry/runtime, non-shared producer paths, and other unproven cases retain the incumbent sparse route. A failure after FIA submission propagates instead of silently replaying the sparse baseline.

`enable_sfa_fia_shared_prefill` is disabled by default, so existing behavior is unchanged unless explicitly enabled.

### Scope and ownership

This is a consumer-only optimization under the existing `skip_topk=True` shared-consumer path. LightningIndexer still computes and publishes TopK exactly as before. This PR does **not** implement partial-row LI bypass, generic whole-FIA routing, or a dynamic profitability model.

### Validation

The final-source semantic oracle covers five decision-relevant cases: Q4096 packed FIA2048 + sparse tail2048; low-fill decline before native submission; two-request grouped packing/restoration (`C=[1024,1024], Q=[1536,1536]`); geometry-plan reuse while consuming the current layer's TopK; and late sparse-tail error propagation with no silent baseline replay or metadata/TopK mutation. All five passed in the frozen source oracle.

Seven focused repository-native tests are included for the default-off/nonshared fallback, Q4096 packing, low-fill fallback, whole-dense/over-cap fallback, two-request ordering, current-TopK plan reuse, and late-tail error/no-mutation contracts.

Existing bounded single-NPU correctness evidence for the same mechanism showed downstream equivalence for Q4096 and the two-request grouped geometry (recorded max absolute difference approximately 0.00390625), with TopK/KV state unchanged. This is not a current-main full-model claim.

A compatible-backport GLM-5.2 TP16 model-lab measurement for Q4096 (`output=1`, concurrency 1, prefix cache off) observed baseline median TTFT 1365.954 ms vs candidate 1151.489 ms, approximately 15.7% lower, with 12/12 paired blocks positive. This is a bounded model-lab signal only; it is not a production or generic-current-main performance claim.

### Non-claims

This PR does not claim production improvement, a universal FIA break-even threshold, or that all full-visible rows should use FIA. The current `dense_total == 2048` admission is a conservative first proven geometry, not a correctness boundary. Unsupported cells fall back to the incumbent sparse consumer.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/attention/sfa_v1.py [hunk 1]
```diff
@@ -33,7 +34,7 @@
     transdata,
     wait_for_kv_layer_from_connector,
 )
-from vllm_ascend.device.device_op import DeviceOperator
+from vllm_ascend.device.device_op import SFA_INDEXER_SPARSE_COUNT, DeviceOperator
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py]: Thanks for catching the invalid import. I reproduced the CI/mypy failure. I did not switch it to `vllm_ascend.device.utils`, because current main does not define `SFA_INDEXER_SPARSE_COUNT` there either, and this PR must remain standalone from the unmerged #15813 branch that introduced a similarly named constant. The fix uses a feature-local `SFA_FIA_SHARED_PREFILL_TOPK_WIDTH = 2048` in `sfa_v1.py`, matching this fast path's current exact TopK/admission geometry and avoiding a cross-PR dependency.

File: tests/ut/attention/test_sfa_v1.py [hunk 2]
```diff
@@ -936,6 +939,205 @@ def setUp(
...
+        metadata.attn_mask = torch.ones(2048, 2048, dtype=torch.int8).triu(1)
+        metadata._sfa_fia_shared_prefill_plan = None
+        indices = torch.arange(total, dtype=torch.int32).view(-1, 1, 1).expand(-1, 1, 2048).contiguous()
+        return q, rope, cache, metadata, indices
+
+    def _mock_shared_fia_kernels(self):
+        fia_patch = patch(
+            "vllm_ascend.attention.sfa_v1.torch_npu.npu_fused_infer_attention_score",
+            side_effect=lambda **kwargs: (kwargs["query"] + 10, None),
+        )
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply tests/ut/attention/test_sfa_v1.py:1078]: Addressed. I added `tests/e2e/pull_request/one_card/test_sfa_v1_shared_prefill.py` with real Ascend kernel registration and candidate-vs-existing-all-sparse coverage for both admitted geometries: single request Q=4096 (FIA 2048 + sparse tail 2048), and grouped C=[1024,1024], Q=[1536,1536] (packed FIA 2048 + sparse tail 1024). The test wraps, rather than substitutes, the real FIA and sparse calls, verifies row restoration, finite/repeat output, TopK/KV/metadata invariance, and fail-closed low-fill/unsupported-block-size cases.

I also bound a direct single-NPU post4 validation to the final production `sfa_v1.py` SHA `cdbd131d1796381776b150ffd9d47485dabd87d02439401e6998bfcb22eea272`: both admitted cases matched the all-sparse reference with max_abs=0.00390625, repeat max_abs=0, and unchanged TopK/KV/block-table/length metadata.

The repository's existing nightly performance infrastructure is model/service-level; there is no operator-level SFA/FIA nightly gate to extend without creating a new framework. Instead this PR includes a bounded non-gating matched benchmark harness with warmup, synchronization, alternating AB/BA visits, raw samples, and no `candidate < baseline` assertion. On the same final source/post4 runtime, component-path medians were 10.2881 ms -> 6.3225 ms for Q4096 and 7.7970 ms -> 3.9867 ms for the grouped case. These are single-NPU component/path measurements, not full-model or production claims.

File: vllm_ascend/ascend_config.py [hunk 3]
```diff
@@ -391,6 +392,9 @@ class AscendConfig:
     enable_mc2_hierarchy_comm: bool = False  # deprecated, will be replaced by mc2_comm_alg = "hierarchy"
     enable_reduce_sample: bool = False
     enable_dsa_cp: bool = False
+    # Experimental post-TopK shared consumer: exactly 2048 dense rows plus a sparse tail.
+    # Default off; supports only the guarded eager split-RoPE FIA configuration.
+    enable_sfa_fia_shared_prefill: bool = False
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/ascend_config.py:459]: Addressed in `docs/source/user_guide/configuration/additional_config.md`. The documentation now lists `enable_sfa_fia_shared_prefill` as an experimental bool (default `False`), includes the `--additional-config` example, records the exact packed-dense-total=2048 plus sparse-tail admission, fixed head/KV/rank/RoPE/block/TopK geometry, eager/runtime/parallel/C8/offload limitations derived from the current source guards, and states that unsupported cases fail closed to the existing sparse consumer. It does not claim generic FIA or production gain.

**Comment by Ruiqiu-Zheng** [comment]
Current HEAD `0f9181ed1c0da5201bdaccd366721f80c5e88d37` has passed the GitHub `pre-commit` job, including the independent mypy step. The review findings above are addressed and the threads have been resolved. When convenient, could a maintainer add `ready-precise` to trigger the selected NPU CI? Thanks.

**Comment by lijiahang226** [comment]
> Current HEAD `0f9181ed1c0da5201bdaccd366721f80c5e88d37` has passed the GitHub `pre-commit` job, including the independent mypy step. The review findings above are addressed and the threads have been resolved. When convenient, could a maintainer add `ready-precise` to trigger the selected NPU CI? Thanks.

Thank you for the contribution. If you'd like to get this PR merged, feel free to join the vLLM Ascend Weekly Meeting on Wednesday, 15:00–16:00 (UTC+8): https://tinyurl.com/vllm-ascend-meeting

**Comment by Ruiqiu-Zheng** [comment]
> > Current HEAD `0f9181ed1c0da5201bdaccd366721f80c5e88d37` has passed the GitHub `pre-commit` job, including the independent mypy step. The review findings above are addressed and the threads have been resolved. When convenient, could a maintainer add `ready-precise` to trigger the selected NPU CI? Thanks.
> 
> Thank you for the contribution. If you'd like to get this PR merged, feel free to join the vLLM Ascend Weekly Meeting on Wednesday, 15:00–16:00 (UTC+8): https://tinyurl.com/vllm-ascend-meeting

Thank you for the invitation and for helping move the PR forward. I really appreciate it. Unfortunately, I won’t be able to join the weekly meeting due to my schedule, but I’ll continue to follow up on the PR and address any feedback promptly. Thanks again.

**Comment by Ruiqiu-Zheng** [comment]
@codex review

**Comment by Ruiqiu-Zheng** [comment]
Current HEAD e2a33a5e7c4fbad7f4e1093b27048d99dc735f73 has completed the selected 310P/A2/A3 CI and ci-gate successfully, and the current-head Codex review found no major issues. This PR is ready for maintainer review when convenient. Thanks.

**Comment by Ruiqiu-Zheng** [comment]
Quick follow-up: the PR has been rebased onto current main and the latest HEAD `dc951c49` is conflict-free. Current-head CI is fully complete with 27 checks passing and no failures. When convenient, could a maintainer take another look? The follow-up PRs are stacked separately, so this parent can be reviewed on its own. Thanks.

File: vllm_ascend/attention/sfa_v1.py [hunk 4]
```diff
@@ -1421,6 +1437,404 @@ def _execute_sparse_flash_attention_process(
...
+            or attn_output.device != ql_nope.device
+        ):
+            raise RuntimeError(
+                "Experimental SFA FIA dense prefill returned an unsupported latent result: "
+                f"expected shape={tuple(ql_nope.shape)}, dtype={ql_nope.dtype}; "
+                f"got shape={tuple(attn_output.shape)}, dtype={attn_output.dtype}, device={attn_output.device}."
+            )
+        return attn_output
+
+    def _try_sfa_fia_shared_prefill(
```
Comment (lijiahang226) [inline vllm_ascend/attention/sfa_v1.py:1771]: Upstream vLLM already has a closely related sparse-MLA optimization: short prefills with seq_len <= topk_tokens are routed to dense MHA, and the indexer can subsequently skip scoring when the MLA route guarantees that the TopK result will not be consumed (#47327 and #48407).

The row-level admission in this PR looks like a reasonable generalization of the same invariant: a query row can use dense attention while its visible KV length is within the TopK width.

My main concern is therefore not the dense-vs-sparse equivalence itself, but where this routing logic is implemented.

Upstream makes dense/sparse prefill routing part of the attention metadata / common MLA execution flow (use_dense_mha), and the backend only executes the selected path. In this PR, _try_sfa_fia_shared_prefill() introduces another planner inside AscendSFAImpl: it analyzes request geometry, partitions and packs rows, caches a plan in generic SFA metadata, constructs synthetic metadata, dispatches FIA/SFA, and restores row order.

Could we align this more closely with the upstream routing abstraction? For example, represent the hybrid prefill decision as an explicit prefill execution plan built with the metadata, and keep AscendSFAImpl responsible for executing that plan rather than discovering and scheduling it inside forward().

This would also avoid relying on exact-class checks such as type(self) is AscendSFAImpl as the mechanism for excluding other SFA implementations.

**Comment by lijiahang226** [comment]
@lrf-vm @Biuapha Could you please take a look at this PR when you have a chance? Your review would be greatly appreciated.

File: vllm_ascend/attention/sfa_v1.py (hunk 4 above)
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:1771]: Thanks for the detailed review and for pointing us to the upstream `use_dense_mha` routing precedent. I refactored this path so the shared SFA/FIA hybrid decision is now represented as an explicit `SFAFIASharedPrefillPlan` built with SFA metadata. `AscendSFAImpl` consumes that plan and keeps runtime validation/execution, instead of rediscovering request geometry and scheduling inside `forward()`. I also removed the exact-class routing check and made unsupported DSA/DCP/KV-offload metadata builders opt out explicitly. The existing grouped row-level FIA admission/execution behavior remains default-off and unchanged; the focused shared-FIA unit tests pass after the refactor. Thanks again — this made the boundary with the existing upstream dense-MHA routing much clearer. Does this direction look aligned with what you had in mind?

**Comment by Ruiqiu-Zheng** [comment]
Current HEAD `6c2307eca91f1d696c7248186be880ed80a553a2` now has the full current-head CI complete with no failures: pre-commit, CPU UT, selected 310P/A2/A3 tests, upstream selected tests, DCO, and ci-gate all passed. The shared-FIA E2E fixture was updated to the metadata-owned `SFAFIASharedPrefillPlan` architecture from the earlier review, without changing the default-off execution semantics. When convenient, could a maintainer take another look? Thanks.

**Comment by Ruiqiu-Zheng** [comment]
Update after the latest main rebase: GitHub now reports the PR mergeable again. The current CPU-UT failure is in two MLA/DCP speculator tests, not the SFA/FIA path. I checked this before changing anything unrelated: the failing speculator/test files have no #16091 diff, and a differential local harness reproduces both failures identically on the current base `5095634` and on #16091 HEAD. The base is the squash of #17336 onto a main that already contained #17206; #17336\x27s own CPU-UT passed on its earlier diverged head, so that CI did not exercise this merged combination. I am therefore not patching unrelated DCP/MLA behavior into this PR. I will keep #16091 scoped to the reviewed metadata-owned SFA/FIA routing and rerun/follow the upstream integration state.

**Comment by Ruiqiu-Zheng** [comment]
Quick follow-up after the upstream DCP/spec-decode fixes landed: the two CPU-UT failures on this PR were reproduced on the then-current upstream base as well, and upstream #16527 has since updated the affected speculator/test path; its CPU-UT and ci-gate are green. #16091 itself remains mergeable and still has `ready-precise`. I do not have repository admin permission to rerun the Actions workflow directly. Could a maintainer please rerun the current PR CI? If it comes back green, could you also confirm whether the metadata-owned `SFAFIASharedPrefillPlan` routing now addresses the earlier architecture concern, and who would be best for the final review? I would like to avoid further code churn unless there is a remaining technical issue. Thanks.

File: vllm_ascend/attention/sfa_v1.py (hunk 4 above)
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:1868]: @lijiahang226 following up specifically on your earlier architecture review here. The planner is now metadata-owned via `SFAFIASharedPrefillPlan`, with `AscendSFAImpl` only consuming the plan, as you suggested. The PR is mergeable again, and the recent CPU-UT blocker was traced to upstream DCP/spec-decode integration and has since been addressed upstream. When you have a chance, could you confirm whether this resolves your original routing-boundary concern? If so, who would be best for the final review? Thanks.
